src/main/java/io/reactivex/internal/operators/flowable/FlowableConcatMap.java

```diff
@@ -13,212 +13,647 @@
 package io.reactivex.internal.operators.flowable;
 
 import java.util.Queue;
-import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.*;
 
 import org.reactivestreams.*;
 
-import io.reactivex.Flowable;
 import io.reactivex.functions.Function;
-import io.reactivex.internal.queue.*;
+import io.reactivex.internal.functions.Objects;
+import io.reactivex.internal.fuseable.QueueSubscription;
+import io.reactivex.internal.queue.SpscArrayQueue;
 import io.reactivex.internal.subscriptions.*;
-import io.reactivex.internal.util.Pow2;
+import io.reactivex.internal.util.Exceptions;
 import io.reactivex.plugins.RxJavaPlugins;
-import io.reactivex.subscribers.SerializedSubscriber;
-
-public final class FlowableConcatMap<T, U> extends Flowable<U> {
-    final Publisher<T> source;
-    final Function<? super T, ? extends Publisher<? extends U>> mapper;
-    final int bufferSize;
-    public FlowableConcatMap(Publisher<T> source, Function<? super T, ? extends Publisher<? extends U>> mapper, int bufferSize) {
-        this.source = source;
-        this.mapper = mapper;
-        this.bufferSize = bufferSize;
+
+public final class FlowableConcatMap<T, R> extends FlowableSource<T, R> {
+final Function<? super T, ? extends Publisher<? extends R>> mapper;
+    
+    final int prefetch;
+    
+    final ErrorMode errorMode;
+    
+    /**
+     * Indicates when an error from the main source should be reported.
+     */
+    public enum ErrorMode {
+        /** Report the error immediately, cancelling the active inner source. */
+        IMMEDIATE,
+        /** Report error after an inner source terminated. */
+        BOUNDARY,
+        /** Report the error after all sources terminated. */
+        END
+    }
+
+    public FlowableConcatMap(Publisher<T> source,
+            Function<? super T, ? extends Publisher<? extends R>> mapper, 
+            int prefetch, ErrorMode errorMode) {
+        super(source);
+        if (prefetch <= 0) {
+            throw new IllegalArgumentException("prefetch > 0 required but it was " + prefetch);
+        }
+        this.mapper = Objects.requireNonNull(mapper, "mapper");
+        this.prefetch = prefetch;
+        this.errorMode = Objects.requireNonNull(errorMode, "errorMode");
+    }
+    
+    public static <T, R> Subscriber<T> subscribe(Subscriber<? super R> s, Function<? super T, ? extends Publisher<? extends R>> mapper, 
+            int prefetch, ErrorMode errorMode) {
+        Subscriber<T> parent = null;
+        switch (errorMode) {
+        case BOUNDARY:
+            parent = new ConcatMapDelayed<T, R>(s, mapper, prefetch, false);
```

Comment:
you can just return from inside these cases and avoid the local var / null initializer problem.